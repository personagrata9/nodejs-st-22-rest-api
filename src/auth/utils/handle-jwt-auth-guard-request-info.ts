import { ForbiddenException, UnauthorizedException } from '@nestjs/common';

export const handleJwtAuthGuardRequestInfo = (info: any) => {
  if (info.message === 'No auth token') {
    throw new UnauthorizedException('authorization required');
  } else {
    throw new ForbiddenException('access denied');
  }
};
