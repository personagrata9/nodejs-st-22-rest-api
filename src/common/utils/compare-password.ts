import * as bcrypt from 'bcrypt';

export const comparePassword = async (
  password: string,
  hash: string,
): Promise<boolean> => {
  return bcrypt.compare(password, hash);
};
