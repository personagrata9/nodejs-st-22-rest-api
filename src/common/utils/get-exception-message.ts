import { HttpException } from '@nestjs/common';
import { IExceptionResponse } from '../interfaces/exception-response.interface';

export const getExceptionMessage = (exception: unknown): string | string[] => {
  if (exception instanceof HttpException) {
    return (exception.getResponse() as IExceptionResponse).message;
  } else if (exception instanceof Error) {
    return exception.message;
  } else {
    return 'Something went wrong!';
  }
};
