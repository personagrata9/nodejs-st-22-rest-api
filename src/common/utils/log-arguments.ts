import { Request, Response } from 'express';
import safeJsonStringify from 'safe-json-stringify';
import { Logger } from 'winston';

export const logArguments = (logger: Logger, req: Request, res: Response) => {
  const reqStringified: string = safeJsonStringify(req, null, 2);
  const resStringified: string = safeJsonStringify(res, null, 2);

  logger.debug(
    `ARGUMENTS\nrequest: ${reqStringified}\n\nresponse: ${resStringified}\n`,
  );
};
