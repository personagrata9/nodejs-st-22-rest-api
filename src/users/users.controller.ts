import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  Param,
  ParseUUIDPipe,
  Post,
  Put,
  Query,
} from '@nestjs/common';
import { CreateUserDto } from './dto/create-user.dto';
import { IUser } from './models/user.model';
import { UsersService } from './services/users.service';
import { UpdateUserDto } from './dto/update-user.dto';
import { UserByIdPipe } from './validation/pipes/user-by-id.pipe';

@Controller('v1/users')
export class UsersController {
  constructor(private readonly usersService: UsersService) {}

  @Get()
  async getAutoSuggestUsers(@Query() query: any) {
    const { limit = 10, offset = 0, loginSubstring } = query;
    return {
      items: await this.usersService.findAutoSuggestUsers(
        +limit,
        +offset,
        loginSubstring,
      ),
      limit: +limit,
      offset: +offset,
      count: await this.usersService.count(),
    };
  }

  @Get(':id')
  async getById(
    @Param('id', new ParseUUIDPipe({ version: '4' }), UserByIdPipe) id: string,
  ): Promise<IUser> {
    return await this.usersService.findOneById(id);
  }

  @Post()
  async create(@Body() createUserDto: CreateUserDto): Promise<IUser> {
    return await this.usersService.create(createUserDto);
  }

  @Put(':id')
  async update(
    @Param('id', new ParseUUIDPipe({ version: '4' }), UserByIdPipe) id: string,
    @Body() updateUserDto: UpdateUserDto,
  ): Promise<IUser> {
    return await this.usersService.update(id, updateUserDto);
  }

  @Delete(':id')
  @HttpCode(204)
  async delete(
    @Param('id', new ParseUUIDPipe({ version: '4' }), UserByIdPipe) id: string,
  ) {
    return await this.usersService.delete(id);
  }
}
