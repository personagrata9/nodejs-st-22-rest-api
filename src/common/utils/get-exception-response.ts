import { HttpException, HttpStatus } from '@nestjs/common';
import { IExceptionResponse } from '../interfaces/exception-response.interface';
import { getExceptionMessage } from './get-exception-message';

export const getExceptionResponse = (
  exception: unknown,
): IExceptionResponse => {
  const statusCode: number =
    exception instanceof HttpException
      ? exception.getStatus()
      : HttpStatus.INTERNAL_SERVER_ERROR;

  const message: string | string[] = getExceptionMessage(exception);

  const error: string =
    exception instanceof HttpException
      ? (exception.getResponse() as IExceptionResponse).error
      : 'Internal Server Error';

  return {
    statusCode,
    message,
    error,
  };
};
