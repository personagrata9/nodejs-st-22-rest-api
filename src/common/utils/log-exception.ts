import { Logger } from 'winston';
import { Request, Response } from 'express';
import { logArguments } from './log-arguments';
import { IExceptionResponse } from '../interfaces/exception-response.interface';

export const logException = (
  logger: Logger,
  exceptionResponse: IExceptionResponse,
  req: Request,
  res: Response,
  context?: string,
) => {
  const { statusCode, message, error } = exceptionResponse;

  const exceptionMessageString =
    typeof message === 'string' ? message : message.join('; ');

  logger.error(`${statusCode} ${error}: ${exceptionMessageString}`, {
    context,
  });

  logArguments(logger, req, res);
};
