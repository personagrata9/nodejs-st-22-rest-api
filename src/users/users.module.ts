import 'dotenv/config';
import { Module } from '@nestjs/common';
import { InMemoryUsersRepository } from './repository/in-memory-users.repository';
import { SequelizeUsersRepository } from './repository/sequelize-users.repository';
import { UsersService } from './services/users.service';
import { UsersController } from './users.controller';

@Module({
  controllers: [UsersController],
  providers: [
    UsersService,
    {
      provide: 'UsersRepository',
      useClass:
        process.env.NODE_ENV === 'test'
          ? InMemoryUsersRepository
          : SequelizeUsersRepository,
    },
  ],
})
export class UsersModule {}
