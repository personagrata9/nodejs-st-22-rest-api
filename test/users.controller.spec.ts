import { Test, TestingModule } from '@nestjs/testing';
import { UsersController } from '../src/users/controllers/users.controller';
import { UsersService } from '../src/users/services/users.service';
import { UsersServiceMock } from './__mocks__/users.service.mock';
import { IUser } from '../src/users/interfaces/user.interface';
import { IPaginatedItemsResult } from '../src/common/interfaces/paginated-items-result.interface';
import { user } from './fixtures/user.fixture';
import { createUserDto } from './fixtures/create-user-dto.fixture';
import { updateUserDto } from './fixtures/update-user-dto.fixture';
import { NotUniqueError } from '../src/common/errors/not-unique.error';
import { BadRequestException } from '@nestjs/common';
import { hashPassword } from '../src/common/utils/hash-password';

describe('UsersController', () => {
  let controller: UsersController;
  let service: any;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [UsersController],
      providers: [
        {
          provide: UsersService,
          useClass: UsersServiceMock,
        },
      ],
    }).compile();

    controller = module.get<UsersController>(UsersController);
    service = module.get<UsersService>(UsersService);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  it('should return array of users with login includes loginSubstring', async () => {
    const loginSubstring: string = user.login.substring(0, 3);
    const limit = 10;
    const offset = 0;

    service.findAutoSuggestUsers.mockResolvedValue({
      items: [user],
      limit,
      offset,
      count: 1,
    });

    const result: IPaginatedItemsResult<IUser> =
      await controller.getAutoSuggestUsers({ limit, offset, loginSubstring });

    expect(result).toEqual({ items: [user], limit, offset, count: 1 });
    expect(service.findAutoSuggestUsers.mock.calls[0]).toMatchSnapshot();
  });

  it('should return user', async () => {
    service.findOneById.mockResolvedValue(user);

    const result: IUser = await controller.getById(user);

    expect(result).toEqual(user);
  });

  it('should create a user', async () => {
    service.create.mockResolvedValue(user);

    const result: IUser = await controller.create(createUserDto);

    expect(result).toEqual(user);
  });

  it('should throw an error when creating a user with not unique login', async () => {
    service.create.mockRejectedValue(
      new NotUniqueError('user', 'login', createUserDto.login),
    );

    try {
      await controller.create(createUserDto);
    } catch (error) {
      expect(error).toBeInstanceOf(BadRequestException);
      expect(error.message).toBe(
        `user with login ${createUserDto.login} already exists, please choose another login`,
      );
    }
  });

  it('should update existing user', async () => {
    const password: string = await hashPassword(updateUserDto.password);
    const updatedUser: IUser = { ...user, ...updateUserDto, password };

    service.update.mockResolvedValue(updatedUser);

    const result: IUser = await controller.update(user, updateUserDto);

    expect(result).toEqual(updatedUser);
  });

  it('should throw an error when updating a user with not unique login', async () => {
    service.update.mockRejectedValue(
      new NotUniqueError('user', 'login', updateUserDto.login),
    );

    try {
      await controller.update(user, updateUserDto);
    } catch (error) {
      expect(error).toBeInstanceOf(BadRequestException);
      expect(error.message).toBe(
        `user with login ${updateUserDto.login} already exists, please choose another login`,
      );
    }
  });

  it('should delete existing user', async () => {
    const result = await controller.delete(user);
    expect(result).toBeUndefined();
  });
});
