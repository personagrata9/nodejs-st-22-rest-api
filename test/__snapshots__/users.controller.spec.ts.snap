// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersController should return array of users with login includes loginSubstring 1`] = `
Array [
  10,
  0,
  "adm",
]
`;
